### PR TITLE
Implement fortune enchantment

### LINKED ISSUE #4800: Fortune Enchant
## Description

I can't add Fortune enchant to item.

## Alternative methods
You should adding the Fortune enchant to the api.

### PR DESCRIPTION
**Pull request by ShockedPlot7560**
### Relevant issues
* Fixes #4800

### API changes
Add the following methods to ``Block``:
- ``getFortuneDrops``: As for the silk touch or the normal drops

Add the ``FortuneTrait`` containing all fortune logics

## Tests

https://youtu.be/m-lTUq6S7fI

### CODE REVIEW

File: src/block/Carrot.php [hunk 1]
```diff
@@ -23,19 +23,42 @@
...
+	}
+
+	public function getFortuneDrops(Item $item) : array{
+		return $this->getDropsForFortuneLevel(
+			$item->getEnchantmentLevel(VanillaEnchantments::FORTUNE())
+		);
+	}
+
+	/** @return Item[] */
+	private function getDropsForFortuneLevel(int $level = 0) : array{
```
Comment (dktapps) [inline src/block/Carrot.php]: Avoid optional parameters. This is a recipe for bugs.

File: src/block/DoubleTallGrass.php [hunk 2]
```diff
@@ -34,9 +33,20 @@ public function canBeReplaced() : bool{
...
 		return [];
 	}
+
+	public function isAffectedByFortune() : bool{
+		return true;
+	}
+
+	public function getFortuneDrops(Item $item) : array{
+		$fortune = VanillaEnchantments::FORTUNE();
+		return $fortune->grassDrops(
```
Comment (dktapps) [inline src/block/DoubleTallGrass.php]: Is this used anywhere outside of this class? I'm not super comfortable with putting specialized methods in general classes.

File: src/block/Leaves.php [hunk 3]
```diff
@@ -173,4 +163,33 @@ public function getFlammability() : int{
...
+	public function getFortuneDrops(Item $item) : array{
+		if(($item->getBlockToolType() & BlockToolType::SHEARS) !== 0){
+			return parent::getDropsForCompatibleTool($item);
+		}
+
+		return $this->getDropsForFortuneLevel($item->getEnchantmentLevel(VanillaEnchantments::FORTUNE()));
+	}
+
+	/** @return Item[] */
+	private function getDropsForFortuneLevel(int $level = 0) : array{
```
Comment (dktapps) [inline src/block/Leaves.php]: Avoid optional parameters.

File: src/block/NetherWartPlant.php [hunk 4]
```diff
@@ -93,7 +96,25 @@ public function onRandomTick() : void{
...
+		$fortuneEnchantment = VanillaEnchantments::FORTUNE();
+		if ($this->age === self::MAX_AGE) {
+			return $fortuneEnchantment->discreteDrops(
+				$this->asItem(),
+				self::MINIMUM_DROPS,
+				self::MAXIMUM_DROPS,
+				$item->getEnchantmentLevel($fortuneEnchantment)
+			);
+		} else {
+			return [$this->asItem()];
```
Comment (dktapps) [inline src/block/NetherWartPlant.php]: I feel like some logic is getting repeated here.

**Review by dktapps — CHANGES_REQUESTED** [review]
I don't think it makes much sense to bake the helpers for this into FortuneEnchantment. It isn't offering any meaningful benefit, since the enchantment is statically referenced, and it's adding a bunch of extra boilerplate code.

File: src/block/Block.php [hunk 5]
```diff
@@ -482,6 +495,13 @@ public function isAffectedBySilkTouch() : bool{
 		return false;
 	}
 
+	/**
+	 * Returns whether Fortune enchanted tools will cause this block to drop more items.
+	 */
+	public function isAffectedByFortune() : bool{
```
Comment (dktapps) [inline src/block/Block.php]: I get the impression this method is probably redundant. If a block isn't affected by fortune, it won't override getFortuneDrops(), and the block's regular drops will be given anyway.

File: src/block/DoubleTallGrass.php (hunk 2 above)
Comment (ShockedPlot7560) [reply src/block/DoubleTallGrass.php]: The drop of wheat seeds is specified in ``DoubleTallGrass`` and ``TallGrass``

**Review by dktapps — COMMENTED** [review]
I'm also getting the impression that there's a lot of duplicated drops logic between fortune and non-fortune. My understanding (from a brief read) is that most of these blocks can reuse the fortune logic for the non-fortune cases, just using a level of 0, and avoid duplicating the regular logic.

Perhaps a trait which redirects `getDropsForCompatibleTool()` to `getFortuneDrops()` (or a helper `getFortuneDropsForLevel()` would simplify the code.

File: src/block/Beetroot.php [hunk 6]
```diff
@@ -23,26 +23,45 @@
...
+			$item->getEnchantmentLevel(VanillaEnchantments::FORTUNE())
+		);
+	}
+
+	/** @return Item[] */
+	private function getDropsForFortuneLevel(int $level = 0) : array{
+		if ($this->age >= self::MAX_AGE) {
+			return array_merge(
+				[VanillaItems::BEETROOT()],
+				VanillaEnchantments::FORTUNE()->binomialDrops(
```
Comment (BrandPVP) [inline src/block/Beetroot.php]: Why isn't this a static method? It will improve performance

Comment (dktapps) [reply src/block/Beetroot.php]: That's a bad reason to make this change, it's a cold path.

**Comment by ShockedPlot7560** [comment]
I'll put all the utilities in a trait dedicated to fortune

**Comment by ShockedPlot7560** [comment]
I just moved all the logic of the fortune into a dedicated line. I'm still mixed on the use of constant for some blocks like ores to specify the min and max drop rate though.

File: src/block/Potato.php [hunk 7]
```diff
@@ -23,23 +23,36 @@
...
+				$level,
+				1
+			);
+		} else {
+			$result = [
+				VanillaItems::POTATO()
+			];
+		}
 		if($this->age >= self::MAX_AGE && mt_rand(0, 49) === 0){
 			$result[] = VanillaItems::POISONOUS_POTATO();
```
Comment (dktapps) [inline src/block/Potato.php:39]: Does fortune have any effect on this?

File: src/block/Wheat.php [hunk 8]
```diff
@@ -23,26 +23,34 @@
...
+	 * @return Item[]
+	 */
+	protected function getFortuneDropsForLevel(int $level) : array{
+		if ($this->age >= self::MAX_AGE) {
+			return array_merge(
+				[VanillaItems::WHEAT()],
+				$this->binomialDrops(
+					VanillaItems::WHEAT_SEEDS(),
+					$level
+				)
```
Comment (dktapps) [inline src/block/Wheat.php]: Perhaps `binomialDrops` could accept `array<Item> $extraDrops` to avoid this array_merge boilerplate everywhere.

File: src/block/utils/FortuneTrait.php [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+					VanillaItems::WHEAT_SEEDS()->setCount($drop)
+				];
+			}
+		}
+
+		return [];
+	}
+
+	/**
+	 * Discret drop, increases the maximum number of items that can be dropped by the fortune level.
```
Comment (dktapps) [inline src/block/utils/FortuneTrait.php]:
```suggestion
	 * Discrete drop increases the maximum number of items that can be dropped by the fortune level.
```

File: src/block/utils/FortuneTrait.php [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+	 *
+	 * @param int $maximumDropLimitation As minecraft doc, this is the maximum number of drops that can be dropped by this enchantment.
+	 *                                   If a drop higher than these maximums is rolled, it is rounded down to the capacity.
+	 * @return Item[]
+	 */
+	protected function discreteDrops(Item $item, int $min, int $max, int $fortuneLevel, int $maximumDropLimitation = PHP_INT_MAX) : array{
```
Comment (dktapps) [inline src/block/utils/FortuneTrait.php]: These parameter names are confusing. `$max` isn't really the max drops, it's the _base_ maximum, to which the fortune level is added. `$maximumDropLimitation` is the real max value.

**Review by dktapps — CHANGES_REQUESTED** [review]
This is looking better, but still needs work.

File: src/block/Potato.php (hunk 7 above)
Comment (ShockedPlot7560) [reply src/block/Potato.php:39]: I don't see any information about it in the wiki. Maybe there is, but since we don't have an exact value, I think we can assume not.

Comment (ShockedPlot7560) [reply src/block/Potato.php:39]: Everywhere it is quoted that the poisoned potatoes are drop with 2% chance but without more

File: src/block/Block.php [hunk 11]
```diff
@@ -566,6 +566,10 @@ public function getDrops(Item $item) : array{
 				return $this->getSilkTouchDrops($item);
 			}
 
+			if ($item->hasEnchantment(VanillaEnchantments::FORTUNE())) {
```
Comment (dktapps) [inline src/block/Block.php]:
```suggestion
			if($item->hasEnchantment(VanillaEnchantments::FORTUNE())){
```

File: src/block/Gravel.php [hunk 12]
```diff
@@ -25,24 +25,33 @@
...
+
+	public function tickFalling() : ?Block{
+		return null;
+	}
+
+	/**
+	 * @return Item[]
+	 */
+	protected function getFortuneDropsForLevel(int $level) : array{
+		if(mt_rand(1, 10 - 3 * $level) === 1){
```
Comment (dktapps) [inline src/block/Gravel.php]: This will crash with fortune 4 or higher

File: src/block/Leaves.php [hunk 13]
```diff
@@ -137,8 +143,42 @@ public function getDropsForCompatibleTool(Item $item) : array{
...
+		);
+	}
+
+	/**
+	 * @return Item[]
+	 */
+	protected function getFortuneDropsForLevel(int $level) : array{
 		$drops = [];
-		if(mt_rand(1, 20) === 1){ //Saplings
+		if(mt_rand(1, 20 - 4 * $level) === 1){ //Saplings
```
Comment (dktapps) [inline src/block/Leaves.php]: This will crash with fortune 5 or higher

File: src/block/Leaves.php [hunk 14]
```diff
@@ -154,29 +194,13 @@ public function getDropsForCompatibleTool(Item $item) : array{
 				$drops[] = $sapling;
 			}
 		}
-		if(($this->leavesType->equals(LeavesType::OAK()) || $this->leavesType->equals(LeavesType::DARK_OAK())) && mt_rand(1, 200) === 1){ //Apples
+		if(($this->leavesType->equals(LeavesType::OAK()) || $this->leavesType->equals(LeavesType::DARK_OAK())) && mt_rand(1, 200 - 20 * $level) === 1){ //Apples
```
Comment (dktapps) [inline src/block/Leaves.php]: This will crash with fortune 10 or higher

File: src/block/Leaves.php [hunk 15]
```diff
@@ -154,29 +194,13 @@ public function getDropsForCompatibleTool(Item $item) : array{
...
 			$drops[] = VanillaItems::APPLE();
 		}
-		if(mt_rand(1, 50) === 1){
+		if(mt_rand(1, 50 - 5 * $level) === 1){
```
Comment (dktapps) [inline src/block/Leaves.php]: Same again here

File: src/block/utils/FortuneTrait.php [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+	 * @return Item[]
+	 */
+	abstract protected function getFortuneDropsForLevel(int $level) : array;
+
+	/**
+	 * Gives a weight of 2 to a normal drop chance and adds a weight of 1 for each extra drop multiplier.
+	 *
+	 * @return Item[]
+	 */
+	protected function weightedDrops(Item $item, int $min, int $max, int $fortuneLevel) : array{
```
Comment (dktapps) [inline src/block/utils/FortuneTrait.php]: A check for min < max should be added here, otherwise this will generate confusing errors on incorrect usage

File: src/block/utils/FortuneTrait.php [hunk 17]
```diff
@@ -59,6 +60,8 @@ abstract protected function getFortuneDropsForLevel(int $level) : array;
 	 * @return Item[]
 	 */
 	protected function weightedDrops(Item $item, int $min, int $max, int $fortuneLevel) : array{
+		assert($min <= $max, "Minimum drop amount must be less than or equal to maximum drop amount");
```
Comment (dktapps) [inline src/block/utils/FortuneTrait.php]: We generally don't use assert for argument validation. This should throw an InvalidArgumentException.

File: src/block/utils/FortuneTrait.php [hunk 18]
```diff
@@ -118,6 +121,8 @@ protected function grassDrops(int $fortuneLevel) : array {
 	 * @return Item[]
 	 */
 	protected function discreteDrops(Item $item, int $minBaseAmount, int $maxBaseAmount, int $fortuneLevel, int $maximum = PHP_INT_MAX) : array{
+		assert($minBaseAmount <= $maxBaseAmount, "Minimum base drop amount must be less than or equal to maximum base drop amount");
```
Comment (dktapps) [inline src/block/utils/FortuneTrait.php]: Same again here.

File: src/block/Gravel.php [hunk 19]
```diff
@@ -25,24 +25,34 @@
...
 
-	public function getDropsForCompatibleTool(Item $item) : array{
-		if(mt_rand(1, 10) === 1){
+	public function isAffectedBySilkTouch() : bool{
+		return true;
+	}
+
+	public function tickFalling() : ?Block{
+		return null;
+	}
```
Comment (dktapps) [inline src/block/Gravel.php]: Unrelated change

File: src/block/Leaves.php [hunk 20]
```diff
@@ -154,29 +195,13 @@ public function getDropsForCompatibleTool(Item $item) : array{
...
-	}
-
-	public function getFlammability() : int{
-		return 60;
-	}
-
-	public function getSupportType(int $facing) : SupportType{
-		return SupportType::NONE();
-	}
 }
```
Comment (dktapps) [inline src/block/Leaves.php]: This diff is quite hard to read due to the logic being moved. I think it would make more sense to have the fortune functions near the original drop functions rather than dumping them at the bottom of every file.

File: src/block/Potato.php [hunk 21]
```diff
@@ -23,16 +23,33 @@
...
 class Potato extends Crops{
+	use FortuneTrait;
 
-	public function getDropsForCompatibleTool(Item $item) : array{
-		$result = [
-			VanillaItems::POTATO()->setCount($this->age >= self::MAX_AGE ? mt_rand(1, 5) : 1)
-		];
+	public function getPickedItem(bool $addUserData = false) : Item{
+		return VanillaItems::POTATO();
+	}
```
Comment (dktapps) [inline src/block/Potato.php]: This doesn't seem related to the PR. It's already handled by asItem().

Again, the diff for this file looks very chaotic due to logic moving around, making it difficult to read.

**Review by dktapps — CHANGES_REQUESTED** [review]
I'd be happier if the fortune drop functions were in the same position in the `Block` classes as the old drops code that they are replacing (or next to them, if they aren't being replaced). Currently, this diff is quite noisy due to all the functions being dumped at the bottom of the files, which is also not great for organization.

File: src/block/Leaves.php [hunk 22]
```diff
@@ -23,22 +23,29 @@
...
+use pocketmine\world\generator\object\TreeType;
 use pocketmine\world\World;
+use function max;
 use function mt_rand;
 
 class Leaves extends Transparent{
+	use FortuneTrait;
+
 	protected LeavesType $leavesType; //immutable for now
+	protected TreeType $treeType;
```
Comment (dktapps) [inline src/block/Leaves.php]: Unrelated change

**Review by dktapps — CHANGES_REQUESTED** [review]

**Comment by ShockedPlot7560** [comment]
It seems that this PR does not implement Sweet berries. If we were to consider doing so, it might break the BC due to the need to add a parameter to the ``getBerryDropAmount`` function for fortune calculation.

File: src/block/Carrot.php [hunk 23]
```diff
@@ -23,15 +23,15 @@
...
 use pocketmine\item\Item;
 use pocketmine\item\VanillaItems;
-use function mt_rand;
 
 class Carrot extends Crops{
 
 	public function getDropsForCompatibleTool(Item $item) : array{
 		return [
-			VanillaItems::CARROT()->setCount($this->age >= self::MAX_AGE ? mt_rand(1, 4) : 1)
+			VanillaItems::CARROT()->setCount($this->age >= self::MAX_AGE ? FortuneDropHelper::binomial($item, 1) : 1)
```
Comment (ShockedPlot7560) [inline src/block/Carrot.php:34]: According to the wiki, it seems that 2 carrots are fixed by default

**Comment by dktapps** [comment]
> It seems that this PR does not implement Sweet berries. If we were to consider doing so, it might break the BC due to the need to add a parameter to the `getBerryDropAmount` function for fortune calculation.

We can't change it due to BC breaks. Really that method shouldn't have been public ...

File: src/block/utils/FortuneDropHelper.php [hunk 24]
```diff
@@ -123,10 +123,7 @@ public static function discrete(Item $usedItem, int $min, int $maxBase, int $max
 			throw new \InvalidArgumentException("Minimum base drop amount must be less than or equal to maximum base drop amount");
 		}
 
-		$max = min(
-			$maxLimit,
-			$maxBase + $usedItem->getEnchantmentLevel(VanillaEnchantments::FORTUNE())
-		);
-		return mt_rand($min, $max);
+		$drop = mt_rand($min, $maxBase + $usedItem->getEnchantmentLevel(VanillaEnchantments::FORTUNE()));
+		return min($drop, $maxLimit);
```
Comment (dktapps) [inline src/block/utils/FortuneDropHelper.php]: I'd actually intended to remove the maxLimit from here.

File: src/block/utils/FortuneDropHelper.php [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+	 *
+	 * @return int the number of items to drop
+	 */
+	public static function discrete(Item $usedItem, int $min, int $maxBase) : int{
+		if($maxBase < $min){
+			throw new \InvalidArgumentException("Minimum base drop amount must be less than or equal to maximum base drop amount");
+		}
+
+		$max = $maxBase + $usedItem->getEnchantmentLevel(VanillaEnchantments::FORTUNE());
+		return mt_rand($min, $max);
```
Comment (ShockedPlot7560) [inline src/block/utils/FortuneDropHelper.php:122]: that's not right. For sealantern, for example, with fortune 3, there's a 20% chance of getting 2, 3, 4, but a 40% chance of getting 5.

Comment (ShockedPlot7560) [reply src/block/utils/FortuneDropHelper.php:122]: This works well for those with a uniform distribution, but not for the rest who have a limit.

Comment (dktapps) [reply src/block/utils/FortuneDropHelper.php:122]: I'm pretty sure it's supposed to work that way.

> Each level of Fortune increases the maximum number of prismarine crystals dropped. **The amount is capped at 5, so Fortune III simply increases the chance of getting 5 crystals.**

Comment (dktapps) [reply src/block/utils/FortuneDropHelper.php:122]: Same story for melons:
> The actual amount is capped at 9, so Fortune III simply increases the chance of getting 9 slices, which is sufficient to craft the melon again from the slices.

For glowstone it's a bit more hazy, but the Java logic for these blocks is identical.

Comment (dktapps) [reply src/block/utils/FortuneDropHelper.php:122]: A quick test case:
```php
$hits = [];

$runs = 10000;

$item = VanillaItems::DIAMOND_PICKAXE()->addEnchantment(new EnchantmentInstance(VanillaEnchantments::FORTUNE(), 3));
for($i = 0; $i < $runs; $i++){
	$amount = min(5, FortuneDropHelper::discrete($item, 2, 3));

	$hits[$amount] = ($hits[$amount] ?? 0) + 1;
}

ksort($hits, SORT_NUMERIC);
foreach($hits as $amount => $count){
	echo "$amount: " . (($count / $runs) * 100) . "%\n";
}
```
Output:
```
2: 20.03%
3: 19.28%
4: 20.68%
5: 40.01%
```

**Comment by dktapps** [comment]
Live on `test.pmmp.io:19132` for testing

File: src/block/utils/FortuneDropHelper.php (hunk 25 above)
Comment (ShockedPlot7560) [reply src/block/utils/FortuneDropHelper.php:122]: > I'm pretty sure it's supposed to work that way.
> 
> > Each level of Fortune increases the maximum number of prismarine crystals dropped. **The amount is capped at 5, so Fortune III simply increases the chance of getting 5 crystals.**

that's surprising because that's not what it says on the wiki fortune's page.

Comment (dktapps) [reply src/block/utils/FortuneDropHelper.php:122]: To paraphrase my Discord comments:

Fortune 3 on sea lantern would, if uncapped, have a 20% chance of dropping 6 crystals. However, it's capped by `min(5, amount)` in `SeaLantern`. This means that 6 gets rounded down to 5. The chance of 5 crystals is then 40%, since there are two possible results which would produce a drop count of 5.

Other blocks have the same behaviour.

File: src/block/Carrot.php (hunk 23 above)
Comment (ShockedPlot7560) [reply src/block/Carrot.php:34]: > The yield is calculated by a binomial distribution: 2 drops are fixed, then a drop is attempted three times with a success rate of 57.14286% to yield the extra 0–3 drops. Each level of Fortune enchantment increases the number of attempts by one.

https://minecraft.fandom.com/wiki/Carrot

Comment (dktapps) [reply src/block/Carrot.php:34]: does this apply to Bedrock?

Comment (dktapps) [reply src/block/Carrot.php:34]: This does not appear to apply to Bedrock from what I can see. Without Fortune I always get 1-4 carrots.

**Review by dktapps — APPROVED** [review]
This now seems good to go.
